### PR TITLE
kvdb: generalize etcd tests

### PR DESCRIPTION
**Pull request by joostjager**
This PR moves the etcd tests one level up so that they can be ran on different backends as well. With that move completed, we add code to run the tests on bolt.

### CODE REVIEW

File: kvdb/bolt_fixture.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		t:       t,
+		tempDir: tempDir,
+	}
+}
+
+func (b *boltFixture) Cleanup() {
+	os.RemoveAll(b.tempDir)
+}
+
+func (b *boltFixture) Check(expected map[string]interface{}) {}
```
Comment (joostjager) [inline kvdb/bolt_fixture.go:32]: To do

**Comment by joostjager** [comment]
@bhandras to make the existing tests pass on bolt, I needed to make changes to the expectations. Especially the cursor change seems to point to an implementation error in etcd, so you may want to take a look at that.

File: kvdb/etcd/bucket_test_funcs.go [hunk 2]
```diff
@@ -4,7 +4,7 @@ package etcd
 
 // bkey is a helper functon used in tests to create a bucket key from passed
 // bucket list.
-func bkey(buckets ...string) string {
+func Bkey(buckets ...string) string {
```
[the remark this answers is not in the document]
Comment (bhandras) [reply kvdb/etcd/bucket_test_funcs.go:7]: Could also just move these functions into `etcd/bucket.go` (I also needed to do this here: https://github.com/lightningnetwork/lnd/pull/5392/commits/51f8175c2e8973c927cfd72392aeb784d3ef27f8)

File: kvdb/readwrite_bucket_test.go [hunk 3]
```diff
@@ -151,15 +151,6 @@ func testBucketDeletion(t *testing.T, db walletdb.DB) {
 		// "apple/pear" deleted
 		require.Nil(t, apple.NestedReadWriteBucket([]byte("pear")))
 
-		// "apple/pear/cherry" deleted
```
Comment (bhandras) [inline kvdb/readwrite_bucket_test.go:181]: This is interesting. Isn't this actually a bbolt bug?

File: kvdb/etcd/readwrite_cursor.go [hunk 4]
```diff
@@ -96,11 +96,6 @@ func (c *readWriteCursor) Prev() (key, value []byte) {
 // not exist, the cursor is moved to the next key after seek.  Returns
 // the new pair.
 func (c *readWriteCursor) Seek(seek []byte) (key, value []byte) {
```
Comment (bhandras) [inline kvdb/etcd/readwrite_cursor.go:98]: nice!

File: kvdb/bolt_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	"testing"
+)
+
+func TestBoltReadCursorEmptyInterval(t *testing.T) {
+	f := NewBoltFixture(t)
+	defer f.Cleanup()
+
+	testReadCursorEmptyInterval(t, f.NewBackend())
+}
+func TestBoltReadCursorNonEmptyInterval(t *testing.T) {
```
Comment (bhandras) [inline kvdb/bolt_test.go:13]: nit: new line

File: kvdb/bolt_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	f := NewBoltFixture(t)
+	defer f.Cleanup()
+
+	testReadCursorNonEmptyInterval(t, f.NewBackend())
+}
+func TestBoltReadWriteCursor(t *testing.T) {
```
Comment (bhandras) [inline kvdb/bolt_test.go:19]: nit: new line

File: kvdb/bolt_test.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	testBucketCreateDelete(t, f.NewBackend())
+}
+
+func TestBoltTxManualCommit(t *testing.T) {
+	f := NewBoltFixture(t)
+	defer f.Cleanup()
+
+	testTxManualCommit(t, f.NewBackend())
+}
+func TestBoltTxRollback(t *testing.T) {
```
Comment (bhandras) [inline kvdb/bolt_test.go:88]: nit: new line

File: kvdb/bolt_test.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
+package kvdb
```
Comment (bhandras) [inline kvdb/bolt_test.go:1]: Scope creep but we could also add a dump and check into these tests.

**Review by bhandras — COMMENTED** [review]
Cool!

File: kvdb/readwrite_bucket_test.go (hunk 3 above)
Comment (joostjager) [reply kvdb/readwrite_bucket_test.go:181]: Bolt crashes hard on it. I think it is just not pointing anywhere anymore.

File: kvdb/bolt_test.go (hunk 8 above)
Comment (joostjager) [reply kvdb/bolt_test.go:1]: Yes considered that, but a dump of bbolt is not that easy, because it needs to go through all buckets recursively. As this test is extra already, left the final assertions out for now.

File: kvdb/etcd/bucket_test_funcs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply kvdb/etcd/bucket_test_funcs.go]: Moved and renamed.

File: kvdb/readwrite_bucket_test.go (hunk 3 above)
Comment (joostjager) [reply kvdb/readwrite_bucket_test.go:181]: Hmm yes maybe. It isn't the most beautiful ending of a process.

File: kvdb/etcd_test.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	vkey = etcd.ValueKey
+)
+
+func TestBucketCreation(t *testing.T) {
+	t.Parallel()
+
+	f := etcd.NewEtcdTestFixture(t)
+	defer f.Cleanup()
+
+	testBucketCreation(t, f.NewBackend())
```
Comment (guggero) [inline kvdb/etcd_test.go:24]: These calls now all look extremely similar, especially when looking at the added `bbolt` tests in a later commit. And since all these test logic functions have the same signature, couldn't we add them to a list and then just loop through them for `etcd` and `bbolt`? Sort of like our itests that we run against different backends. Could just skip the `require.Equal(t, expected, f.Dump())` part for `bbolt`...

**Review by guggero — APPROVED** [review]
Awesome that we can now run these tests against `bbolt` (and later Postgres) almost "for free" :100: 
And it seems we already found and fixed a discrepancy between `etcd` and `bbolt` which is great!

Just one comment about making the tests a bit more generic, but otherwise LGTM :tada:

File: kvdb/etcd_test.go (hunk 9 above)
Comment (joostjager) [reply kvdb/etcd_test.go]: Done

File: kvdb/etcd_test.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+			},
+		},
+		{
+			name:       "tx rollback",
+			test:       testTxRollback,
+			expectedDb: map[string]string{},
+		},
+	}
+
+	for _, test := range tests {
```
Comment (bhandras) [inline kvdb/etcd_test.go:137]: nit: The linter doesn't complain here because the test is behind a tag, but I believe we need `test := test` added.

File: kvdb/bolt_test.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+			test: testTxManualCommit,
+		},
+		{
+			name: "tx rollback",
+			test: testTxRollback,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
```
Comment (bhandras) [inline kvdb/bolt_test.go:69]: hmm looks like linter doesn't care

**Review by bhandras — APPROVED** [review]
LGTM :100:

File: kvdb/etcd_test.go (hunk 10 above)
Comment (joostjager) [reply kvdb/etcd_test.go:137]: Added, also for bbolt.

This is such a nasty bug always. And I think things were going wrong here too.

File: kvdb/bolt_test.go (hunk 11 above)
Comment (joostjager) [reply kvdb/bolt_test.go:71]: indeed

**Comment by joostjager** [comment]
@bhandras I re-requested review because I found issues with the test assertions, see last commit.
